ayeletK/assignment3/grid.cpp

```diff
@@ -0,0 +1,120 @@
+#include "position.cpp"
+#include "dictionary.cpp"
+
+class Grid{
+    int _rows, _cols;
+    vector<vector<char>> _grid;
+    bool update_location(int* row, int* col){
+        if (*col + 1 < _cols){ 
+            (*col)++;
+            return true;
+        }
+        else if (*row + 1 < _rows) {
+            *col = 0;
+            (*row)++;
+            return true;
+        }
+        return false;
+    }
+    bool legal_row(int row){
+        return row >=0 && row < _rows;
+    }
+    bool legal_col(int col){
+        return col >=0 && col < _cols;
+    }
+    bool legal_position(Position pos){
+        return legal_row(pos._row) && legal_col(pos._col);
+    }
+    set<Position> get_neighbours(Position curr_pos){
+        Position neighbour;
+        set<Position> neighbours;
+        for(int row=curr_pos._row - 1; row <= curr_pos._row + 1; ++row){
+            for(int col=curr_pos._col - 1; col <= curr_pos._col + 1; ++col){
+                neighbour = Position(row, col);
+                if (legal_position(neighbour) && !curr_pos.equal(neighbour)) {
+                    neighbours.insert(neighbour);
+                }
+            }
+        }
+        return neighbours;
+    }
+    set<Position> get_unvisited_neighbours(Position curr_pos, set<Position>* visited){
+        set<Position> res;
+        for (Position pos: get_neighbours(curr_pos)){
+            if (visited->find(pos) == visited->end())
+                res.insert(pos);
+        }
+        return res;
+    }
+    void words_with_prefix(Position curr_pos, set<Position>* visited, 
+    Dictionary* dict, vector<char>* prefix, set<string>* words){
+        prefix->push_back(get_value(curr_pos));
+        visited->insert(curr_pos);
+        string curr_prefix = string(prefix->begin(), prefix->end());
+        if (! dict->is_prefix(curr_prefix)) {
+            visited->erase(curr_pos);
+            prefix->pop_back();
+            return;
+        }
+        if (dict->is_word(curr_prefix))
+            words->insert(curr_prefix);
+        for (Position neighbour: get_unvisited_neighbours(curr_pos, visited))
+            words_with_prefix(neighbour, visited, dict, prefix, words);
+        visited->erase(curr_pos);
+        prefix->pop_back();
+    }
+    set<string> find_words_from_position(Position start_pos, Dictionary* dict){
+        set<string> res;
+        vector<char> prefix;
+        set<Position> visited;  // can be improved to hashset ?
```

Comment:
In case the order of the elements in the set is not important, then a hash set (e.g. `std::unordered_set`) can be more efficient than a tree set (e.g. `std::set`).
However the difference between the sets it is not that important for me in the scope of this course, as they are both pretty efficient. As long as you make the correct logical use of sets, both C++ library sets are OK (and easily interchangeable too).

Reply:
Noted. i will leave it in a set this time.